govuk_chat_evaluation/output_guardrails/evaluate.py

```
@@ -28,15 +34,125 @@ def classification(self) -> str:
+
+    @staticmethod
+    def parse_exact_string(exact_string: str, num_guardrails: int) -> tuple[bool, str]:
+        """Parses the exact guardrail result string for guardrail triggers
+
+        Returns:
+            A tuple containing:
+                - A boolean indicating if any guardrail was triggered.
+                - A string of comma-separated guardrail numbers if triggered,
+                  otherwise the string "None".
+        """
+        # Pattern distinguishes between quoted comma-separated numbers and the literal None
+        pattern = r"^(True|False)\s*\|\s*(?:\"(.*)\"|(None))$"
+        match = re.match(pattern, exact_string)
+
+        if not match:
+            raise ValueError(f"Incorrect format for guardrail result: {exact_string}")
+
+        triggered_str = match.group(1)
+        guardrails_str = match.group(2)  # comma-separated guardrailnumbers, if present
+        none_str = match.group(3)  # The literal string "None", if present
+
+        if triggered_str == "True":
+            if guardrails_str is None:
+                raise ValueError(
+                    f"Incorrect format for guardrail result: {exact_string}"
+                )
+
+            # Check for empty or whitespace guardrailsstring
+            if not guardrails_str.strip():
+                raise ValueError(
+                    f"Incorrect format for guardrail result: {exact_string}"
+                )
+
+            guardrail_numbers = [int(num) for num in guardrails_str.split(",")]
+            # Check guardrail numbers are in range and unique
+            if not all(1 <= num <= num_guardrails for num in guardrail_numbers):
+                raise ValueError(
+                    f"Incorrect format for guardrail result: {exact_string}"
+                )
+            if len(guardrail_numbers) != len(set(guardrail_numbers)):
+                raise ValueError(
+                    f"Incorrect format for guardrail result: {exact_string}"
+                )
```

Question: are we too strict here?

**First case - some out-of-range predicted indexes.**
If the LLM guardrail returns "True | "1,3,9", we currently discard the entire result because the "9" is outside the valid range. Instead of discarding the entire result, we could consider raising a warning for the invalid guardrail (i.e., "9"), but still proceed with the valid indices (i.e., "1" and "3").

**Second case - duplicate indeces.**
If the LLM guardrail returns "True | "1,3,3", we are currently discarding the entire result due to the duplicate "3". An alternative approach could be to raise a warning for the duplicate value and then proceed by de-duplicating the list of indices before continuing with the evaluation.

---

Yes, good point. I should distinguish between things that should give a warning, but can still be evaluated, and things that need to raise an error.